File: truezip-driver/truezip-driver-tzp/src/main/java/de/truezip/driver/zip/raes/LazyOutputSocket.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.truezip.driver.zip.raes;

import de.schlichtherle.truezip.cio.Entry;
import de.schlichtherle.truezip.io.DecoratingOutputStream;
import de.schlichtherle.truezip.cio.DecoratingOutputSocket;
import de.schlichtherle.truezip.cio.OutputSocket;
import edu.umd.cs.findbugs.annotations.CreatesObligation;
import java.io.IOException;
import java.io.OutputStream;
import javax.annotation.concurrent.NotThreadSafe;

/**
 * A lazy output socket provides proxy output streams which acquire their
 * underlying local target upon the first read access.
 *
 * @param  <E> the type of the {@link #getLocalTarget() local target}.
 * @see    LazyInputSocket
 * @author Christian Schlichtherle
 * @deprecated This class will be removed in TrueZIP 8.
 */
@Deprecated
@NotThreadSafe
final class LazyOutputSocket<E extends Entry>
extends DecoratingOutputSocket<E> {

    LazyOutputSocket(OutputSocket<? extends E> output) {
        super(output);
    }

    /**
     * Returns a proxy output stream which acquires its underlying output
     * stream upon the first read access.
     *
     * @return A proxy output stream which acquires its underlying output
     *         stream upon the first write access.
     */
    @Override
    public OutputStream newOutputStream() {
        return new ProxyOutputStream();
    }

    @NotThreadSafe
    private class ProxyOutputStream extends DecoratingOutputStream {
        @CreatesObligation
        @edu.umd.cs.findbugs.annotations.SuppressWarnings("OBL_UNSATISFIED_OBLIGATION")
        ProxyOutputStream() {
            super(null);
        }

        OutputStream getDelegate() throws IOException {
            final OutputStream out = delegate;
            return null != out
                    ? out
                    : (delegate = getBoundDelegate().newOutputStream());
        }

        @Override
        public void write(int b) throws IOException {
            getDelegate().write(b);
        }

        @Override
        public void write(byte[] b, int off, int len) throws IOException {
            getDelegate().write(b, off, len);
        }

        @Override
        public void flush() throws IOException {
            getDelegate().flush();
        }

        @Override
        public void close() throws IOException {
            final OutputStream out = delegate;
            if (null != out)
                out.close();
        }
    } // ProxyOutputStream
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/LazyOutputSocket.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel

import javax.annotation.concurrent._
import net.truevfs.kernel.cio._

/** An output socket which obtains its delegate socket lazily and `reset`s
  * it upon any [[java.lang.Throwable]].
  *
  * @see    LazyInputSocket
  * @tparam E the type of the `target`.
  * @author Christian Schlichtherle
  */
@NotThreadSafe
private abstract class LazyOutputSocket[E <: Entry]
extends AbstractOutputSocket[E] with LazySocketLike[E, OutputSocket[_ <: E]] {
  override def target() = apply(_.target())
  override def stream(peer: AnyInputSocket) = apply(_.stream(peer))
  override def channel(peer: AnyInputSocket) = apply(_.channel(peer))
}
